### PR TITLE
Per-group web UI

### PR DESCRIPTION
**Pull request by retroinspect**
- Add view by group feature to operator & frontend
- `localhost:1443?groupId={groupId}` now shows devices only belong to group of `groupId`

### CODE REVIEW

File: frontend/src/liboperator/operator/devices.go [hunk 1]
```diff
@@ -93,12 +107,41 @@ func (d *Device) ToClient(id int, msg interface{}) error {
...
+}
+
+func GetGroupId(info interface{}) string {
+	deviceInfo, ok := info.(map[string]interface{})
+	if !ok {
+		return DEFAULT_GROUP_ID
+	}
+
+	groupId, ok := deviceInfo["group_id"].(string)
+	if !ok {
```
Comment (ikicha) [inline frontend/src/liboperator/operator/devices.go:127]: if !ok || len(groupId) == 0?

File: frontend/src/liboperator/operator/devices.go [hunk 2]
```diff
@@ -93,12 +107,41 @@ func (d *Device) ToClient(id int, msg interface{}) error {
...
+
+	if len(groupId) == 0 {
+		return DEFAULT_GROUP_ID
+	}
+
+	return groupId
+}
+
+func GetGroup(p *DevicePool, groupId string) *Group {
+	_, ok := p.groups[groupId]
```
Comment (ikicha) [inline frontend/src/liboperator/operator/devices.go:139]: is it okay to return nil if the groupId is not found? if so, we can just return p.groups[groupId]

File: frontend/src/liboperator/operator/devices.go [hunk 3]
```diff
@@ -93,12 +107,41 @@ func (d *Device) ToClient(id int, msg interface{}) error {
...
+	}
+
+	if len(groupId) == 0 {
+		return DEFAULT_GROUP_ID
+	}
+
+	return groupId
+}
+
+func GetGroup(p *DevicePool, groupId string) *Group {
```
Comment (ikicha) [inline frontend/src/liboperator/operator/devices.go:138]: It is actually CreateOrGetGroup, please rename it.

File: frontend/src/liboperator/operator/devices.go [hunk 4]
```diff
@@ -138,3 +211,28 @@ func (p *DevicePool) DeviceIds() []string {
...
+	p.devicesMtx.Lock()
+	defer p.devicesMtx.Unlock()
+	ret := make([]*DeviceInfo, 0)
+	for id, device := range p.devices {
+		ret = append(ret, &DeviceInfo{DeviceId: id, GroupId: device.groupId})
+	}
+	return ret
+}
+
+func (p *DevicePool) GetDeviceInfoListByGroupId(groupId string) []*DeviceInfo {
```
Comment (ikicha) [inline frontend/src/liboperator/operator/devices.go:225]: can we add GetDeviceInfoListByPredicate or something, and let GetDeviceInfoList and GetDeviceInfoListByGroupId use it?

File: frontend/src/operator/webui/src/app/device.service.ts [hunk 5]
```diff
@@ -1,37 +1,59 @@
...
 
-  private devices = merge(
-    this.deviceFromServer,
-    this.refreshSubject.pipe(mergeMap(() => this.deviceFromServer))
-  ).pipe(shareReplay(1));
+  private devices = this.groupIdSubject.pipe(
+    combineLatestWith(this.refreshSubject),
+    mergeMap(([groupId, _]) => {
+      if (groupId === null) {
+        return this.allDeviceFromServer;
```
Comment (ikicha) [inline frontend/src/operator/webui/src/app/device.service.ts:21]: does refresh actually work for non-group scene?

File: frontend/src/liboperator/operator/devices.go (hunk 4 above)
Comment (retroinspect) [reply frontend/src/liboperator/operator/devices.go:225]: To do that I think GetDeviceInfoList should be first refactored to refer group map of the pool(p.group) to get all devices:

```
func (p *DevicePool) GetDeviceInfoList() []*DeviceInfo {
	p.devicesMtx.Lock()
	defer p.devicesMtx.Unlock()
	ret := make([]*DeviceInfo, 0)

	for groupId, group := range p.groups {
		for _, deviceId := range group.deviceIds {
			ret = append(ret, &DeviceInfo{DeviceId: deviceId, GroupId: groupId})
		}
	}
	return ret
}
```

The only common part between GetDeviceInfoList and GetDeviceInfoListByGroupId becomes the following snippet,
thus GetDeviceInfoListByPredicate do the following:
```
	for _, deviceId := range group.deviceIds {
		ret = append(ret, &DeviceInfo{DeviceId: deviceId, GroupId: groupId})
	}
```

Is this what you mean by?

Comment (ikicha) [reply frontend/src/liboperator/operator/devices.go:221]: nvm, I just thought we could get deviceIds from p.devices with predicates, but for group, the current code looks better.

File: frontend/src/operator/webui/src/app/device.service.ts (hunk 5 above)
Comment (retroinspect) [reply frontend/src/operator/webui/src/app/device.service.ts:21]: I guess "non-group scene" means the view with all devices. 
Refresh works for the case.

**Review by ikicha — APPROVED** [review]

**Pull request merged**